### PR TITLE
Expose c10 cuda ops to caffe2

### PR DESCRIPTION
**Pull request by smessmer**
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#18036 Expose c10 cuda ops to caffe2**&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D14467495/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #18037 Specify overload name in function schema&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D14467497/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #18038 [wip] Allow registering same operator schema multiple times&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D14467494/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #18090 Move schema inference to c10&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D14491454/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #18051 [wip] Revamp c10 kernel registration&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D14475195/)

- Add macros to export c10 cuda operators to caffe2 frontend
- Instead of having a separate caffe2 registry for the c10 operator wrappers, use the existing caffe2 registries

Differential Revision: [D14467495](https://our.internmc.facebook.com/intern/diff/D14467495/)

### CODE REVIEW

**Comment by smessmer** [comment]
this is a re-export of https://github.com/pytorch/pytorch/pull/17827.

File: aten/src/ATen/hip/impl/HIPGuardImplMasqueradingAsCUDA.cpp [hunk 1]
```diff
@@ -12,3 +12,4 @@
 // This hack can be removed once PyTorch is out-of-place HIPified, and
 // doesn't pretend CUDA is HIP.
 C10_REGISTER_GUARD_IMPL(CUDA, at::cuda::HIPGuardImplMasqueradingAsCUDA);
+C10_REGISTER_GUARD_IMPL(HIP, at::cuda::HIPGuardImplMasqueradingAsCUDA);
```
Comment (ezyang) [inline aten/src/ATen/hip/impl/HIPGuardImplMasqueradingAsCUDA.cpp]: This looks wrong. Even in Torch, if I access something referring to it by its true HIP name, I had better get something that actually reports itself to be HIP, and not lying to say it's CUDA.

**Review by ezyang — APPROVED** [review]
approved with comments, it looks like ROCm is broken

**Comment by ezyang** [comment]
```
21:05:45 ======================================================================
21:05:45 ERROR: test_streams (test_cuda.TestCuda)
21:05:45 ----------------------------------------------------------------------
21:05:45 Traceback (most recent call last):
21:05:45   File "/var/lib/jenkins/workspace/test/common_utils.py", line 298, in wrapper
21:05:45     method(*args, **kwargs)
21:05:45   File "/var/lib/jenkins/workspace/test/test_cuda.py", line 1571, in test_streams
21:05:45     self.assertTrue(user_stream.query())
21:05:45   File "/var/lib/jenkins/.local/lib/python2.7/site-packages/torch/cuda/streams.py", line 75, in query
21:05:45     return super(Stream, self).query()
21:05:45 RuntimeError: d.type() == DeviceType::CUDA ASSERT FAILED at /var/lib/jenkins/workspace/aten/src/ATen/hip/impl/HIPGuardImplMasqueradingAsCUDA.h:64, please report a bug to PyTorch.
21:05:45 
21:05:45 ----------------------------------------------------------------------
```

File: aten/src/ATen/hip/impl/HIPGuardImplMasqueradingAsCUDA.cpp (hunk 1 above)
Comment (smessmer) [reply aten/src/ATen/hip/impl/HIPGuardImplMasqueradingAsCUDA.cpp]: This was an attempt to solve a rocm CI issue, but it doesn't seem to have helped. cc @bddppq

**Pull request closed without merging**